src/AppInstallerRepositoryCore/InstallerMetadataCollectionContext.cpp

```diff
@@ -0,0 +1,900 @@
+                    AICLI_LOG(Repo, Error, << "Don't know how to output for version " << m_inputVersion.ToString());
+                    THROW_HR(HRESULT_FROM_WIN32(ERROR_UNSUPPORTED_TYPE));
+                }
+            }
+            catch (...)
+            {
+                CollectErrorDataFromException(std::current_exception());
+            }
+        }
+
+        if (ContainsError())
+        {
+            // We only have one version currently
+            outputJSON = CreateErrorJson_1_0();
+        }
+
+        // Write output
+        outputJSON.serialize(output);
+    }
+
+    std::unique_ptr<ThreadLocalStorage::PreviousThreadGlobals> InstallerMetadataCollectionContext::InitializeLogging(const std::filesystem::path& logFile)
+    {
+        auto threadGlobalsLifetime = m_threadGlobals.SetForCurrentThread();
+
+        Logging::Log().SetLevel(Logging::Level::Info);
+        Logging::Log().EnableChannel(Logging::Channel::All);
+        Logging::EnableWilFailureTelemetry();
+        Logging::AddTraceLogger();
+
+        if (!logFile.empty())
+        {
+            Logging::AddFileLogger(logFile);
+        }
+
+        Logging::Telemetry().SetCaller("installer-metadata-collection");
+        Logging::Telemetry().LogStartup();
+
+        return threadGlobalsLifetime;
+    }
+
+    void InstallerMetadataCollectionContext::InitializePreinstallState(const std::wstring& json)
+    {
+        try
+        {
+            AICLI_LOG(Repo, Info, << "Parsing input JSON:\n" << ConvertToUTF8(json));
+
+            // Parse and validate JSON
+            try
+            {
+                utility::string_t versionFieldName = L"version";
+
+                web::json::value inputValue = web::json::value::parse(json);
+
+                THROW_HR_IF(APPINSTALLER_CLI_ERROR_JSON_INVALID_FILE, inputValue.is_null());
+
+                m_inputVersion = Version{ GetRequiredString(inputValue, versionFieldName) };
+                AICLI_LOG(Repo, Info, << "Parsing input JSON version " << m_inputVersion.ToString());
+
+                if (m_inputVersion.PartAt(0).Integer == 1)
+                {
+                    // We only have one version currently, so use that as long as the major version is 1
+                    ParseInputJson_1_0(inputValue);
+                }
+                else
+                {
+                    AICLI_LOG(Repo, Error, << "Don't know how to handle version " << m_inputVersion.ToString());
+                    THROW_HR(HRESULT_FROM_WIN32(ERROR_UNSUPPORTED_TYPE));
+                }
+            }
+            catch (const web::json::json_exception& exc)
+            {
+                AICLI_LOG(Repo, Error, << "Exception parsing input JSON: " << exc.what());
+                throw;
+            }
+
+            // Collect pre-install system state
+            m_correlationData->CapturePreInstallSnapshot();
+        }
+        catch (...)
+        {
+            CollectErrorDataFromException(std::current_exception());

```

Comment:
This is cool. I didn't know `std::current_exception()` existed.

Reply:
It's great for making standardized exception handlers.